src/utility/animation.ts

```diff
@@ -0,0 +1,67 @@
+export class AnimationFrame {
+    constructor(
+        public x: number,
+        public y: number,
+        public w: number,
+        public h: number) {}
+}
+
+export class AnimationEngine {
+    private loop = false;
+    private frameIndex = 0;
+    private frameTimer = 0;
+    private frameTimeout = 0;
+    private currentState = "";
+
+    constructor(
+        private states: Record<string, AnimationFrame[]>,
+        fps: number
+    ) {
+        this.frameTimeout = 1 / fps;
+    }
```

Comment:
Make the state type safe by making the `AnimationEngine` generic:

```ts
export class AnimationEngine<StateKey extends string = string>. {
    private loop = false;
    private frameIndex = 0;
    private frameTimer = 0;
    private frameTimeout = 0;
    private currentState: StateKey; // <-- type safe

    constructor(
        private states: Record<T, AnimationFrame[]>,
        fps: number
        // probably provide the default state here
        // ...
    ) {
        // and optionally use the first state as the default
        // ...
        this.frameTimeout = 1 / fps;
    }
```
so you can later set the state in a type-safe way:
```ts
   setState(name: StateKey, looping = false)
```

Reply:
Btw you don't need to provide the StateKey type explicitly, typescript will infer it from the `states` type automatically.